src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs

```diff
@@ -0,0 +1,318 @@
+                {
+                    // Pipe consumer is shut down, do we stop writing
+                    break;
+                }
+            }
+        }
+
+        private void FireStreamClosed()
+        {
+            ThreadPool.UnsafeQueueUserWorkItem(state =>
+            {
+                state.CancelConnectionClosedToken();
+
+                state._waitForConnectionClosedTcs.TrySetResult(null);
+            },
+            this,
+            preferLocal: false);
+        }
+
+        private void CancelConnectionClosedToken()
+        {
+            try
+            {
+                _streamClosedTokenSource.Cancel();
+            }
+            catch (Exception ex)
+            {
+                _log.LogError(0, ex, $"Unexpected exception in {nameof(QuicStreamContext)}.{nameof(CancelConnectionClosedToken)}.");
+            }
+        }
+
+
+        private async Task DoSend()
+        {
+            Exception shutdownReason = null;
+            Exception unexpectedError = null;
+
+            try
+            {
+                await ProcessSends();
+            }
+            catch (QuicException ex)
+            {
+                shutdownReason = new ConnectionResetException(ex.Message, ex);
+            }
+            catch (Exception ex)
+            {
+                shutdownReason = ex;
+                unexpectedError = ex;
+                _log.ConnectionError(ConnectionId, unexpectedError);
+            }
+            finally
+            {
+                await ShutdownWrite(shutdownReason);
+
+                // Complete the output after disposing the stream
+                Output.Complete(unexpectedError);
+
+                // Cancel any pending flushes so that the input loop is un-paused
+                Input.CancelPendingFlush();
+            }
+        }
+
+        private async Task ProcessSends()
+        {
+            // Resolve `output` PipeReader via the IDuplexPipe interface prior to loop start for performance.
+            var output = Output;
+            while (true)
+            {
+                var result = await output.ReadAsync();
+
+                if (result.IsCanceled)
+                {
+                    break;
+                }
+
+                var buffer = result.Buffer;
+
+                var end = buffer.End;
+                var isCompleted = result.IsCompleted;
+                if (!buffer.IsEmpty)
+                {
+                    await _stream.WriteAsync(buffer, endStream: isCompleted);
+                }
+
+                output.AdvanceTo(end);
+
+                if (isCompleted)
+                {
+                    // Once the stream pipe is closed, shutdown the stream.
+                    break;
+                }
+            }
+        }
+
+        public override void Abort(ConnectionAbortedException abortReason)
+        {
+            // Don't call _stream.Shutdown and _stream.Abort at the same time.
```

Comment:
Is it OK to call _stream.Abort after _stream.Shutdown?

Reply:
Yep.